File: python-version/LC3356/min_zero_array.py
class Solution:
    def minZeroArray(self, nums: list[int], queries: list[list[int]]) -> int:
        q = len(queries)
        low, high = 0, q+1
        while low < high:
            mid = low + (high - low) // 2
            if self.__check(nums, queries, mid):
                high = mid
            else:
                low = mid + 1
        return low if low <= q else -1

    @staticmethod
    def __check(nums: list[int], queries: list[list[int]], k: int) -> bool:
        n = len(nums)
        diff_array = [0] * n
        for query in queries[:k]:
            diff_array[query[0]] += query[2]
            if query[1] < n - 1:
                diff_array[query[1] + 1] -= query[2]

        sum_diff = 0
        for i, diff in enumerate(diff_array):
            sum_diff += diff
            if nums[i] > sum_diff:
                return False
        return True
        

File: go-version/leetcode/LC3356/min_zero_array.go
package LC3356

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/5/22 上午12:13
 * @FilePath: min_zero_array
 * @Description:
 */

func minZeroArray(nums []int, queries [][]int) int {
	queryLength := len(queries)
	low, high := 0, queryLength+1
	for low < high {
		mid := low + (high-low)/2
		if check(nums, queries, mid) {
			high = mid
		} else {
			low = mid + 1
		}
	}
	if low <= queryLength {
		return low
	}
	return -1
}

func check(nums []int, queries [][]int, k int) bool {
	n := len(nums)
	diffArray := make([]int, n)
	for _, query := range queries[:k] {
		diffArray[query[0]] += query[2]
		if query[1] < n-1 {
			diffArray[query[1]+1] -= query[2]
		}
	}

	sumDiff := 0
	for i, diff := range diffArray {
		sumDiff += diff
		if nums[i] > sumDiff {
			return false
		}
	}
	return true
}
